hypershift-operator/controllers/scheduler/dedicated_content_serving_nodes.go

```diff
@@ -0,0 +1,204 @@
+	log := ctrl.LoggerFrom(ctx, "controller", "DedicatedServingComponentScheduler", "hostedcluster", req.NamespacedName.String())
+	err := r.Get(ctx, req.NamespacedName, hcluster)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("hostedcluster not found, aborting reconcile", "name", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get cluster %q: %w", req.NamespacedName, err)
+	}
+	if !hcluster.DeletionTimestamp.IsZero() {
+		log.Info("hostedcluster is deleted, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	if hcTopology := hcluster.Annotations[hyperv1.TopologyAnnotation]; hcTopology != hyperv1.DedicatedRequestServingComponentsTopology {
+		log.Info("hostedcluster does not use isolated request serving components, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	if scheduled := hcluster.Annotations[hyperv1.HostedClusterScheduledAnnotation]; scheduled == "true" {
+		log.Info("hostedcluster is already scheduled, nothing to do")
+		return ctrl.Result{}, nil
+	}
+	nodeList := &corev1.NodeList{}
+	if err := r.List(ctx, nodeList, client.HasLabels{hyperv1.RequestServingComponentLabel}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	nodesToUse := map[string]*corev1.Node{}
+	// first, find any existing nodes already labeled for this hostedcluster
+	for i := range nodeList.Items {
+		node := &nodeList.Items[i]
+		zone, hasZoneLabel := node.Labels["topology.kubernetes.io/zone"]
+		if !hasZoneLabel {
+			continue
+		}
+		hcLabel, hasHCLabel := node.Labels[hyperv1.HostedClusterLabel]
+		if !hasHCLabel {
+			continue
+		}
+		if hcLabel == fmt.Sprintf("%s-%s", hcluster.Namespace, hcluster.Name) {
+			nodesToUse[zone] = node
+			log.Info("Found existing node for hosted cluster", "node", node.Name, "zone", zone)
+		}
+	}
+
+	if len(nodesToUse) < 2 {
+		for i := range nodeList.Items {
+			node := &nodeList.Items[i]
+			zone, hasZoneLabel := node.Labels["topology.kubernetes.io/zone"]
+			if !hasZoneLabel {
+				// No zone has been set on the node, we cannot use it
+				continue
+			}
+			_, hasHCLabel := node.Labels[hyperv1.HostedClusterLabel]
+			if hasHCLabel {
+				// The node has been allocated to a different hosted cluster, skip it
+				continue
+			}
+			if nodesToUse[zone] == nil {
+				log.Info("Found node to allocate for hosted cluster", "node", node.Name, "zone", zone)
+				nodesToUse[zone] = node
+			}
+			if len(nodesToUse) == 2 {
+				break
+			}
+		}
+	}
+	if len(nodesToUse) < 2 {
+		return ctrl.Result{}, fmt.Errorf("failed to find enough available nodes for cluster, found %d", len(nodesToUse))
+	}
+	for _, node := range nodesToUse {
+		originalNode := node.DeepCopy()
+
+		// Add taint and labels for specific hosted cluster
+		node.Spec.Taints = append(node.Spec.Taints, corev1.Taint{
+			Key:    HostedClusterTaint,
+			Value:  fmt.Sprintf("%s-%s", hcluster.Namespace, hcluster.Name),
+			Effect: corev1.TaintEffectNoSchedule,
+		})
```

Comment:
if we previously found nodes already assigned to the cluster, then they will most likely have this taint already, and we will add the taint again to the list anyway.
This should not cause any problems, but the taints list will keep growing with the same duplicate taint.

Reply:
Thanks, will fix.